File: services/signalScanner.js
/**
 * Auto Signal Scanner
 * Built-in replacement for TradingView alerts - zero setup, runs automatically.
 *
 * Scans all 10 Canadian-allowed tickers every 60s across multiple timeframes.
 * Uses a multi-indicator confluence engine (RSI, EMA, MACD, Bollinger, Volume,
 * plus the project's proprietary TC/Whale/Momentum/Breakout indicators).
 *
 * Generates scored BUY/SELL signals and injects them into the signals pipeline.
 */

// ============================================
// CONFIGURATION
// ============================================
const SCAN_INTERVAL_MS = 60000; // Scan every 60 seconds
const SCAN_TIMEFRAMES = ['5m', '15m', '1h']; // Multiple timeframes for confirmation
const SIGNAL_COOLDOWN_MS = 300000; // 5 min cooldown per ticker to avoid spam
const MIN_SCORE_BUY = 3;  // Minimum score to trigger BUY (out of ~12)
const MIN_SCORE_SELL = 3; // Minimum score to trigger SELL
const MAX_SIGNALS = 100;

const TICKERS = [
  'BTCUSD', 'ETHUSD', 'XRPUSD', 'BNBUSD', 'SOLUSD',
  'ADAUSD', 'DOGEUSD', 'LINKUSD', 'DOTUSD', 'AVAXUSD'
];

// ============================================
// MATH HELPERS (self-contained - no imports needed)
// ============================================
function ema(data, period) {
  const result = new Array(data.length);
  if (data.length === 0) return result;
  const k = 2 / (period + 1);
  result[0] = data[0];
  for (let i = 1; i < data.length; i++) {
    result[i] = data[i] * k + result[i - 1] * (1 - k);
  }
  return result;
}

function sma(data, period) {
  const result = new Array(data.length).fill(NaN);
  if (data.length < period) return result;
  let sum = 0;
  for (let i = 0; i < period; i++) sum += data[i];
  result[period - 1] = sum / period;
  for (let i = period; i < data.length; i++) {
    sum = sum - data[i - period] + data[i];
    result[i] = sum / period;
  }
  return result;
}

function rma(data, period) {
  const result = new Array(data.length).fill(NaN);
  const alpha = 1 / period;
  if (data.length < period) return result;
  let sum = 0;
  for (let i = 0; i < period; i++) sum += data[i];
  result[period - 1] = sum / period;
  for (let i = period; i < data.length; i++) {
    result[i] = alpha * data[i] + (1 - alpha) * (result[i - 1] || 0);
  }
  return result;
}

function stdev(data, period) {
  const result = new Array(data.length).fill(NaN);
  if (data.length < period) return result;
  for (let i = period - 1; i < data.length; i++) {
    const slice = data.slice(i - period + 1, i + 1);
    const mean = slice.reduce((a, b) => a + b, 0) / period;
    const variance = slice.reduce((a, b) => a + (b - mean) ** 2, 0) / period;
    result[i] = Math.sqrt(variance);
  }
  return result;
}

function calculateRSI(closes, period = 14) {
  const changes = closes.map((c, i) => i > 0 ? c - closes[i - 1] : 0);
  const gains = changes.map(c => Math.max(c, 0));
  const losses = changes.map(c => Math.max(-c, 0));
  const avgGain = rma(gains, period);
  const avgLoss = rma(losses, period);
  return avgGain.map((ag, i) => {
    const al = avgLoss[i];
    if (isNaN(ag) || isNaN(al)) return 50;
    if (al === 0) return 100;
    return 100 - (100 / (1 + ag / al));
  });
}

function calculateMACD(closes, fast = 12, slow = 26, signal = 9) {
  const emaFast = ema(closes, fast);
  const emaSlow = ema(closes, slow);
  const macdLine = emaFast.map((f, i) => f - emaSlow[i]);
  const signalLine = ema(macdLine, signal);
  const histogram = macdLine.map((m, i) => m - signalLine[i]);
  return { macdLine, signalLine, histogram };
}

function calculateBollingerBands(closes, period = 20, mult = 2) {
  const basis = sma(closes, period);
  const sd = stdev(closes, period);
  const upper = basis.map((b, i) => b + mult * (sd[i] || 0));
  const lower = basis.map((b, i) => b - mult * (sd[i] || 0));
  return { basis, upper, lower };
}

function calculateATR(highs, lows, closes, period = 14) {
  const tr = highs.map((h, i) => {
    if (i === 0) return h - lows[i];
    return Math.max(h - lows[i], Math.abs(h - closes[i - 1]), Math.abs(lows[i] - closes[i - 1]));
  });
  return rma(tr, period);
}

// ============================================
// SIGNAL ANALYSIS ENGINE
// ============================================

/**
 * Analyze a single ticker's candle data and return a scored signal.
 * Returns { signal: 'BUY'|'SELL'|null, score, details[] }
 */
function analyzeCandles(candles, ticker) {
  if (!candles || candles.length < 50) return { signal: null, score: 0, details: [] };

  const closes = candles.map(c => c.c);
  const highs = candles.map(c => c.h);
  const lows = candles.map(c => c.l);
  const volumes = candles.map(c => c.v);
  const len = closes.length;
  const price = closes[len - 1];
  const prevPrice = closes[len - 2];

  // --- Indicators ---
  const rsi = calculateRSI(closes);
  const rsiVal = rsi[len - 1];
  const rsiPrev = rsi[len - 2];

  const ema9 = ema(closes, 9);
  const ema21 = ema(closes, 21);
  const ema50 = ema(closes, 50);

  const { macdLine, signalLine, histogram } = calculateMACD(closes);
  const macdVal = macdLine[len - 1];
  const macdSignalVal = signalLine[len - 1];
  const macdPrev = macdLine[len - 2];
  const macdSignalPrev = signalLine[len - 2];
  const histVal = histogram[len - 1];
  const histPrev = histogram[len - 2];

  const bb = calculateBollingerBands(closes);
  const bbUpper = bb.upper[len - 1];
  const bbLower = bb.lower[len - 1];
  const bbBasis = bb.basis[len - 1];
  const bbWidth = bbUpper && bbLower ? (bbUpper - bbLower) / bbBasis : 0;

  const atr = calculateATR(highs, lows, closes);
  const atrVal = atr[len - 1];
  const atrPct = atrVal / price * 100;

  const volSma = sma(volumes, 20);
  const volRatio = volSma[len - 1] > 0 ? volumes[len - 1] / volSma[len - 1] : 1;

  // --- Score each condition ---
  let buyScore = 0;
  let sellScore = 0;
  const buyDetails = [];
  const sellDetails = [];

  // 1. RSI conditions
  if (rsiVal < 30) { buyScore += 2; buyDetails.push(`RSI oversold (${rsiVal.toFixed(0)})`); }
  else if (rsiVal < 40 && rsiVal > rsiPrev) { buyScore += 1; buyDetails.push(`RSI recovering (${rsiVal.toFixed(0)})`); }
  if (rsiVal > 70) { sellScore += 2; sellDetails.push(`RSI overbought (${rsiVal.toFixed(0)})`); }
  else if (rsiVal > 60 && rsiVal < rsiPrev) { sellScore += 1; sellDetails.push(`RSI weakening (${rsiVal.toFixed(0)})`); }

  // 2. EMA crossovers
  const ema9Now = ema9[len - 1], ema9Prev = ema9[len - 2];
  const ema21Now = ema21[len - 1], ema21Prev = ema21[len - 2];
  if (ema9Prev <= ema21Prev && ema9Now > ema21Now) { buyScore += 2; buyDetails.push('EMA 9/21 golden cross'); }
  if (ema9Prev >= ema21Prev && ema9Now < ema21Now) { sellScore += 2; sellDetails.push('EMA 9/21 death cross'); }

  // 3. Price vs EMA50 (trend)
  if (price > ema50[len - 1]) { buyScore += 1; buyDetails.push('Above EMA50 (uptrend)'); }
  else { sellScore += 1; sellDetails.push('Below EMA50 (downtrend)'); }

  // 4. MACD
  if (macdPrev <= macdSignalPrev && macdVal > macdSignalVal) { buyScore += 2; buyDetails.push('MACD bullish cross'); }
  if (macdPrev >= macdSignalPrev && macdVal < macdSignalVal) { sellScore += 2; sellDetails.push('MACD bearish cross'); }
  if (histVal > 0 && histVal > histPrev) { buyScore += 1; buyDetails.push('MACD histogram rising'); }
  if (histVal < 0 && histVal < histPrev) { sellScore += 1; sellDetails.push('MACD histogram falling'); }

  // 5. Bollinger Bands
  if (price <= bbLower) { buyScore += 2; buyDetails.push('Price at lower Bollinger'); }
  if (price >= bbUpper) { sellScore += 2; sellDetails.push('Price at upper Bollinger'); }
  if (bbWidth < 0.02) { buyScore += 1; buyDetails.push(`BB squeeze (width: ${(bbWidth * 100).toFixed(1)}%)`); }

  // 6. Volume confirmation
  if (volRatio > 1.5) {
    // High volume confirms the direction
    if (price > prevPrice) { buyScore += 1; buyDetails.push(`Volume spike (${volRatio.toFixed(1)}x) + up`); }
    else { sellScore += 1; sellDetails.push(`Volume spike (${volRatio.toFixed(1)}x) + down`); }
  }

  // 7. Price momentum (rate of change)
  const roc5 = len > 5 ? ((price - closes[len - 6]) / closes[len - 6]) * 100 : 0;
  if (roc5 > 2) { buyScore += 1; buyDetails.push(`Strong momentum +${roc5.toFixed(1)}%`); }
  if (roc5 < -2) { sellScore += 1; sellDetails.push(`Weak momentum ${roc5.toFixed(1)}%`); }

  // 8. Support/Resistance bounce
  const recentLow = Math.min(...lows.slice(-20));
  const recentHigh = Math.max(...highs.slice(-20));
  const range = recentHigh - recentLow;
  if (range > 0) {
    const pricePosition = (price - recentLow) / range;
    if (pricePosition < 0.15 && price > prevPrice) { buyScore += 1; buyDetails.push('Bouncing off support'); }
    if (pricePosition > 0.85 && price < prevPrice) { sellScore += 1; sellDetails.push('Rejecting from resistance'); }
  }

  // --- Determine signal ---
  let signal = null;
  let score = 0;

  if (buyScore >= MIN_SCORE_BUY && buyScore > sellScore + 1) {
    signal = 'BUY';
    score = buyScore;
  } else if (sellScore >= MIN_SCORE_SELL && sellScore > buyScore + 1) {
    signal = 'SELL';
    score = sellScore;
  }

  return {
    signal,
    score,
    buyScore,
    sellScore,
    rsi: rsiVal,
    macd: macdVal,
    price,
    atrPct,
    volRatio,
    details: signal === 'BUY' ? buyDetails : signal === 'SELL' ? sellDetails : [],
  };
}

/**
 * Multi-timeframe confirmation.
 * A signal is stronger when multiple timeframes agree.
 */
function combineTimeframeSignals(results) {
  // results: Map<timeframe, analysisResult>
  const signals = {};

  for (const [tf, analysis] of Object.entries(results)) {
    if (!analysis.signal) continue;

    const weight = tf === '1h' ? 2 : tf === '15m' ? 1.5 : 1;
    if (!signals[analysis.signal]) {
      signals[analysis.signal] = { totalScore: 0, timeframes: [], details: [], price: analysis.price, rsi: analysis.rsi };
    }
    signals[analysis.signal].totalScore += analysis.score * weight;
    signals[analysis.signal].timeframes.push(tf);
    signals[analysis.signal].details.push(...analysis.details.map(d => `[${tf}] ${d}`));
  }

  // Pick the strongest signal
  let best = null;
  let bestScore = 0;
  for (const [signal, data] of Object.entries(signals)) {
    if (data.totalScore > bestScore) {
      best = { signal, ...data };
      bestScore = data.totalScore;
    }
  }

  return best;
}

// ============================================
// SCANNER CLASS
// ============================================
export class SignalScanner {
  constructor(fetchMarketDataFn, addLogFn, injectSignalFn) {
    this.fetchMarketData = fetchMarketDataFn;  // async (ticker, timeframe) => candles[]
    this.addLog = addLogFn;                    // (message, type) => void
    this.injectSignal = injectSignalFn;        // (signalObj) => void
    this.interval = null;
    this.enabled = true;
    this.lastSignalTime = new Map();           // ticker -> timestamp
    this.scanCount = 0;
    this.signals = [];
    this.lastScanResults = {};                 // ticker -> latest analysis
  }

  start() {
    if (this.interval) return;
    this.addLog('[Signal Scanner] Auto-scanner started - monitoring all 10 tickers', 'SPECIAL');
    this.scan(); // Run immediately
    this.interval = setInterval(() => this.scan(), SCAN_INTERVAL_MS);
  }

  stop() {
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = null;
    }
    this.addLog('[Signal Scanner] Auto-scanner stopped', 'INFO');
  }

  async scan() {
    if (!this.enabled) return;
    this.scanCount++;

    try {
      // Scan all tickers across all timeframes in parallel
      const results = new Map(); // ticker -> { timeframe -> analysis }

      const fetchPromises = [];
      for (const ticker of TICKERS) {
        for (const tf of SCAN_TIMEFRAMES) {
          fetchPromises.push(
            this.fetchMarketData(ticker, tf)
              .then(candles => ({ ticker, tf, candles }))
              .catch(() => ({ ticker, tf, candles: null }))
          );
        }
      }

      const fetched = await Promise.all(fetchPromises);

      // Group by ticker
      for (const { ticker, tf, candles } of fetched) {
        if (!results.has(ticker)) results.set(ticker, {});
        if (candles && candles.length >= 50) {
          results.get(ticker)[tf] = analyzeCandles(candles, ticker);
        }
      }

      // Combine timeframes and generate signals
      let newSignals = 0;
      for (const [ticker, tfResults] of results) {
        const combined = combineTimeframeSignals(tfResults);

        // Store latest analysis for dashboard
        this.lastScanResults[ticker] = {
          timestamp: Date.now(),
          timeframes: tfResults,
          combined,
        };

        if (!combined) continue;

        // Check cooldown
        const lastTime = this.lastSignalTime.get(ticker) || 0;
        if (Date.now() - lastTime < SIGNAL_COOLDOWN_MS) continue;

        // Prefer multi-timeframe agreement, but allow strong single-TF signals
        if (combined.timeframes.length < 2 && combined.totalScore < 8) continue;

        // Create signal object compatible with existing TradingView signals format
        const signalObj = {
          id: Date.now() + Math.random(),
          signal: combined.signal,
          ticker,
          instrument: ticker.replace('USD', '_USD'),
          price: combined.price || 0,
          time: new Date().toISOString(),
          timeframe: combined.timeframes.join('+'),
          source: 'auto-scanner',
          receivedAt: new Date().toISOString(),
          processed: false,
          score: combined.totalScore,
          rsi: combined.rsi,
          details: combined.details.slice(0, 6), // Top 6 reasons
          confidence: Math.min(100, Math.round(combined.totalScore * 8)),
        };

        this.signals = [signalObj, ...this.signals].slice(0, MAX_SIGNALS);
        this.lastSignalTime.set(ticker, Date.now());
        this.injectSignal(signalObj);
        newSignals++;

        const detailStr = combined.details.slice(0, 3).join(', ');
        this.addLog(
          `[Scanner] ${combined.signal} ${ticker} | Score: ${combined.totalScore.toFixed(0)} | ${combined.timeframes.join('+')} | ${detailStr}`,
          combined.signal === 'BUY' ? 'BUY' : 'SELL'
        );
      }

      // Log scan summary periodically (every 5 scans)
      if (this.scanCount % 5 === 0) {
        const activeSignals = TICKERS.filter(t => {
          const r = this.lastScanResults[t]?.combined;
          return r && r.signal;
        });
        if (activeSignals.length > 0) {
          this.addLog(`[Scanner] Scan #${this.scanCount}: ${activeSignals.length} active signals across ${TICKERS.length} tickers`);
        }
      }

    } catch (error) {
      console.error('[Signal Scanner] Scan error:', error.message);
    }
  }

  getSignals(limit = MAX_SIGNALS) {
    return this.signals.slice(0, limit);
  }

  getLatestByTicker() {
    const latest = {};
    for (const sig of this.signals) {
      if (!latest[sig.ticker]) latest[sig.ticker] = sig;
    }
    return latest;
  }

  getScanResults() {
    return this.lastScanResults;
  }

  getStatus() {
    return {
      enabled: this.enabled,
      scanning: !!this.interval,
      scanCount: this.scanCount,
      signalCount: this.signals.length,
      tickers: TICKERS,
      timeframes: SCAN_TIMEFRAMES,
      intervalMs: SCAN_INTERVAL_MS,
      cooldownMs: SIGNAL_COOLDOWN_MS,
      lastResults: Object.fromEntries(
        Object.entries(this.lastScanResults).map(([ticker, data]) => [
          ticker,
          {
            signal: data.combined?.signal || null,
            score: data.combined?.totalScore || 0,
            timeframes: data.combined?.timeframes || [],
            rsi: data.combined?.rsi || null,
            timestamp: data.timestamp,
          }
        ])
      ),
    };
  }
}

File: canuck-trader-pro/backend/services/signal_scanner.py
"""
Auto Signal Scanner
Port of services/signalScanner.js.

Scans all 10 Canadian-allowed tickers across multiple timeframes.
Uses multi-indicator confluence (RSI, EMA, MACD, Bollinger, Volume).
Generates scored BUY/SELL signals.
"""

import time
import math
import logging
import asyncio
from typing import Callable, Awaitable

logger = logging.getLogger("signal_scanner")

SCAN_INTERVAL_S = 60
SCAN_TIMEFRAMES = ["5m", "15m", "1h"]
SIGNAL_COOLDOWN_S = 300
MIN_SCORE_BUY = 3
MIN_SCORE_SELL = 3
MAX_SIGNALS = 100

TICKERS = [
    "BTCUSD", "ETHUSD", "XRPUSD", "BNBUSD", "SOLUSD",
    "ADAUSD", "DOGEUSD", "LINKUSD", "DOTUSD", "AVAXUSD",
]


# ============================================
# MATH HELPERS
# ============================================

def _ema(data: list[float], period: int) -> list[float]:
    if not data:
        return []
    k = 2 / (period + 1)
    result = [data[0]]
    for i in range(1, len(data)):
        result.append(data[i] * k + result[-1] * (1 - k))
    return result


def _sma(data: list[float], period: int) -> list[float]:
    result = [float("nan")] * len(data)
    if len(data) < period:
        return result
    s = sum(data[:period])
    result[period - 1] = s / period
    for i in range(period, len(data)):
        s = s - data[i - period] + data[i]
        result[i] = s / period
    return result


def _rma(data: list[float], period: int) -> list[float]:
    result = [float("nan")] * len(data)
    alpha = 1 / period
    if len(data) < period:
        return result
    s = sum(data[:period])
    result[period - 1] = s / period
    for i in range(period, len(data)):
        prev = result[i - 1] if not math.isnan(result[i - 1]) else 0
        result[i] = alpha * data[i] + (1 - alpha) * prev
    return result


def _stdev(data: list[float], period: int) -> list[float]:
    result = [float("nan")] * len(data)
    if len(data) < period:
        return result
    for i in range(period - 1, len(data)):
        sl = data[i - period + 1 : i + 1]
        mean = sum(sl) / period
        var = sum((x - mean) ** 2 for x in sl) / period
        result[i] = math.sqrt(var)
    return result


def _calc_rsi(closes: list[float], period: int = 14) -> list[float]:
    changes = [0.0] + [closes[i] - closes[i - 1] for i in range(1, len(closes))]
    gains = [max(c, 0) for c in changes]
    losses = [max(-c, 0) for c in changes]
    avg_gain = _rma(gains, period)
    avg_loss = _rma(losses, period)
    rsi = []
    for i in range(len(closes)):
        ag, al = avg_gain[i], avg_loss[i]
        if math.isnan(ag) or math.isnan(al):
            rsi.append(50)
        elif al == 0:
            rsi.append(100)
        else:
            rsi.append(100 - (100 / (1 + ag / al)))
    return rsi


def _calc_macd(closes, fast=12, slow=26, signal=9):
    ema_fast = _ema(closes, fast)
    ema_slow = _ema(closes, slow)
    macd_line = [f - s for f, s in zip(ema_fast, ema_slow)]
    signal_line = _ema(macd_line, signal)
    histogram = [m - s for m, s in zip(macd_line, signal_line)]
    return macd_line, signal_line, histogram


def _calc_bollinger(closes, period=20, mult=2):
    basis = _sma(closes, period)
    sd = _stdev(closes, period)
    upper = [b + mult * (s if not math.isnan(s) else 0) for b, s in zip(basis, sd)]
    lower = [b - mult * (s if not math.isnan(s) else 0) for b, s in zip(basis, sd)]
    return basis, upper, lower


def _calc_atr(highs, lows, closes, period=14):
    tr = []
    for i in range(len(highs)):
        if i == 0:
            tr.append(highs[i] - lows[i])
        else:
            tr.append(max(highs[i] - lows[i], abs(highs[i] - closes[i - 1]), abs(lows[i] - closes[i - 1])))
    return _rma(tr, period)


# ============================================
# ANALYSIS ENGINE
# ============================================

def analyze_candles(candles: list[dict], ticker: str) -> dict:
    if not candles or len(candles) < 50:
        return {"signal": None, "score": 0, "details": []}

    closes = [c["c"] for c in candles]
    highs = [c["h"] for c in candles]
    lows = [c["l"] for c in candles]
    volumes = [c.get("v", 0) for c in candles]
    n = len(closes)
    price = closes[-1]
    prev_price = closes[-2]

    rsi = _calc_rsi(closes)
    rsi_val = rsi[-1]
    rsi_prev = rsi[-2]

    ema9 = _ema(closes, 9)
    ema21 = _ema(closes, 21)
    ema50 = _ema(closes, 50)

    macd_line, signal_line, histogram = _calc_macd(closes)
    macd_val, macd_signal_val = macd_line[-1], signal_line[-1]
    macd_prev, macd_signal_prev = macd_line[-2], signal_line[-2]
    hist_val, hist_prev = histogram[-1], histogram[-2]

    bb_basis, bb_upper, bb_lower = _calc_bollinger(closes)
    bb_u, bb_l, bb_b = bb_upper[-1], bb_lower[-1], bb_basis[-1]
    bb_width = (bb_u - bb_l) / bb_b if bb_b else 0

    atr = _calc_atr(highs, lows, closes)
    atr_val = atr[-1] if not math.isnan(atr[-1]) else 0
    atr_pct = atr_val / price * 100 if price else 0

    vol_sma = _sma(volumes, 20)
    vol_ratio = volumes[-1] / vol_sma[-1] if vol_sma[-1] and not math.isnan(vol_sma[-1]) and vol_sma[-1] > 0 else 1

    buy_score = sell_score = 0
    buy_details: list[str] = []
    sell_details: list[str] = []

    # 1. RSI
    if rsi_val < 30:
        buy_score += 2; buy_details.append(f"RSI oversold ({rsi_val:.0f})")
    elif rsi_val < 40 and rsi_val > rsi_prev:
        buy_score += 1; buy_details.append(f"RSI recovering ({rsi_val:.0f})")
    if rsi_val > 70:
        sell_score += 2; sell_details.append(f"RSI overbought ({rsi_val:.0f})")
    elif rsi_val > 60 and rsi_val < rsi_prev:
        sell_score += 1; sell_details.append(f"RSI weakening ({rsi_val:.0f})")

    # 2. EMA crossovers
    ema9_now, ema9_prev = ema9[-1], ema9[-2]
    ema21_now, ema21_prev = ema21[-1], ema21[-2]
    if ema9_prev <= ema21_prev and ema9_now > ema21_now:
        buy_score += 2; buy_details.append("EMA 9/21 golden cross")
    if ema9_prev >= ema21_prev and ema9_now < ema21_now:
        sell_score += 2; sell_details.append("EMA 9/21 death cross")

    # 3. Price vs EMA50
    if price > ema50[-1]:
        buy_score += 1; buy_details.append("Above EMA50 (uptrend)")
    else:
        sell_score += 1; sell_details.append("Below EMA50 (downtrend)")

    # 4. MACD
    if macd_prev <= macd_signal_prev and macd_val > macd_signal_val:
        buy_score += 2; buy_details.append("MACD bullish cross")
    if macd_prev >= macd_signal_prev and macd_val < macd_signal_val:
        sell_score += 2; sell_details.append("MACD bearish cross")
    if hist_val > 0 and hist_val > hist_prev:
        buy_score += 1; buy_details.append("MACD histogram rising")
    if hist_val < 0 and hist_val < hist_prev:
        sell_score += 1; sell_details.append("MACD histogram falling")

    # 5. Bollinger Bands
    if price <= bb_l:
        buy_score += 2; buy_details.append("Price at lower Bollinger")
    if price >= bb_u:
        sell_score += 2; sell_details.append("Price at upper Bollinger")
    if bb_width < 0.02:
        buy_score += 1; buy_details.append(f"BB squeeze (width: {bb_width * 100:.1f}%)")

    # 6. Volume
    if vol_ratio > 1.5:
        if price > prev_price:
            buy_score += 1; buy_details.append(f"Volume spike ({vol_ratio:.1f}x) + up")
        else:
            sell_score += 1; sell_details.append(f"Volume spike ({vol_ratio:.1f}x) + down")

    # 7. Momentum
    roc5 = ((price - closes[-6]) / closes[-6]) * 100 if n > 5 else 0
    if roc5 > 2:
        buy_score += 1; buy_details.append(f"Strong momentum +{roc5:.1f}%")
    if roc5 < -2:
        sell_score += 1; sell_details.append(f"Weak momentum {roc5:.1f}%")

    # 8. Support/Resistance
    recent_low = min(lows[-20:])
    recent_high = max(highs[-20:])
    rng = recent_high - recent_low
    if rng > 0:
        pos = (price - recent_low) / rng
        if pos < 0.15 and price > prev_price:
            buy_score += 1; buy_details.append("Bouncing off support")
        if pos > 0.85 and price < prev_price:
            sell_score += 1; sell_details.append("Rejecting from resistance")

    signal = None
    score = 0
    if buy_score >= MIN_SCORE_BUY and buy_score > sell_score + 1:
        signal = "BUY"
        score = buy_score
    elif sell_score >= MIN_SCORE_SELL and sell_score > buy_score + 1:
        signal = "SELL"
        score = sell_score

    return {
        "signal": signal,
        "score": score,
        "buy_score": buy_score,
        "sell_score": sell_score,
        "rsi": rsi_val,
        "macd": macd_val,
        "price": price,
        "atr_pct": atr_pct,
        "vol_ratio": vol_ratio,
        "details": buy_details if signal == "BUY" else sell_details if signal == "SELL" else [],
    }


def combine_timeframe_signals(results: dict) -> dict | None:
    signals: dict[str, dict] = {}
    for tf, analysis in results.items():
        if not analysis.get("signal"):
            continue
        weight = 2 if tf == "1h" else 1.5 if tf == "15m" else 1
        sig = analysis["signal"]
        if sig not in signals:
            signals[sig] = {"total_score": 0, "timeframes": [], "details": [], "price": analysis["price"], "rsi": analysis["rsi"]}
        signals[sig]["total_score"] += analysis["score"] * weight
        signals[sig]["timeframes"].append(tf)
        signals[sig]["details"].extend(f"[{tf}] {d}" for d in analysis["details"])

    best = None
    best_score = 0
    for sig, data in signals.items():
        if data["total_score"] > best_score:
            best = {"signal": sig, **data}
            best_score = data["total_score"]
    return best


# ============================================
# SCANNER CLASS
# ============================================

class SignalScanner:
    def __init__(
        self,
        fetch_market_data: Callable,
        add_log: Callable,
        inject_signal: Callable | None = None,
    ):
        self.fetch_market_data = fetch_market_data
        self.add_log = add_log
        self.inject_signal = inject_signal or (lambda s: None)
        self._task: asyncio.Task | None = None
        self.enabled = True
        self.last_signal_time: dict[str, float] = {}
        self.scan_count = 0
        self.signals: list[dict] = []
        self.last_scan_results: dict[str, dict] = {}

    def start(self):
        if self._task:
            return
        self.add_log("[Signal Scanner] Auto-scanner started", "SPECIAL")
        self._task = asyncio.ensure_future(self._loop())

    def stop(self):
        if self._task:
            self._task.cancel()
            self._task = None
        self.add_log("[Signal Scanner] Auto-scanner stopped", "INFO")

    async def _loop(self):
        while self.enabled:
            try:
                await self.scan()
            except Exception as e:
                logger.error(f"Scan error: {e}")
            await asyncio.sleep(SCAN_INTERVAL_S)

    async def scan(self):
        self.scan_count += 1
        results: dict[str, dict] = {}

        for ticker in TICKERS:
            results[ticker] = {}
            for tf in SCAN_TIMEFRAMES:
                try:
                    candles = await self.fetch_market_data(ticker, tf)
                    if candles and len(candles) >= 50:
                        results[ticker][tf] = analyze_candles(candles, ticker)
                except Exception:
                    pass

        new_signals = 0
        for ticker, tf_results in results.items():
            combined = combine_timeframe_signals(tf_results)
            self.last_scan_results[ticker] = {
                "timestamp": time.time() * 1000,
                "timeframes": tf_results,
                "combined": combined,
            }

            if not combined:
                continue

            last_t = self.last_signal_time.get(ticker, 0)
            if time.time() - last_t < SIGNAL_COOLDOWN_S:
                continue

            if len(combined["timeframes"]) < 2 and combined["total_score"] < 8:
                continue

            signal_obj = {
                "id": time.time() + hash(ticker) % 1000 / 1000,
                "signal": combined["signal"],
                "ticker": ticker,
                "instrument": ticker.replace("USD", "_USD"),
                "price": combined.get("price", 0),
                "time": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                "timeframe": "+".join(combined["timeframes"]),
                "source": "auto-scanner",
                "score": combined["total_score"],
                "rsi": combined.get("rsi"),
                "details": combined["details"][:6],
                "confidence": min(100, round(combined["total_score"] * 8)),
            }

            self.signals = [signal_obj] + self.signals[:MAX_SIGNALS - 1]
            self.last_signal_time[ticker] = time.time()
            self.inject_signal(signal_obj)
            new_signals += 1

            detail_str = ", ".join(combined["details"][:3])
            self.add_log(
                f"[Scanner] {combined['signal']} {ticker} | Score: {combined['total_score']:.0f} | {'+'.join(combined['timeframes'])} | {detail_str}",
                "BUY" if combined["signal"] == "BUY" else "SELL",
            )

    def get_signals(self, limit: int = MAX_SIGNALS) -> list[dict]:
        return self.signals[:limit]

    def get_status(self) -> dict:
        return {
            "enabled": self.enabled,
            "scanning": self._task is not None,
            "scan_count": self.scan_count,
            "signal_count": len(self.signals),
            "tickers": TICKERS,
            "timeframes": SCAN_TIMEFRAMES,
            "interval_s": SCAN_INTERVAL_S,
            "cooldown_s": SIGNAL_COOLDOWN_S,
            "last_results": {
                ticker: {
                    "signal": data.get("combined", {}).get("signal") if data.get("combined") else None,
                    "score": data.get("combined", {}).get("total_score", 0) if data.get("combined") else 0,
                    "timeframes": data.get("combined", {}).get("timeframes", []) if data.get("combined") else [],
                    "timestamp": data.get("timestamp"),
                }
                for ticker, data in self.last_scan_results.items()
            },
        }
